### ISSUE TITLE
Wagtail Tags prefetching issue creates n+1 problem

### ISSUE THREAD

**Issue by Redhaus** [wagtail/wagtail#6044]
I am not sure if this is a bug or a limitation within Wagtail, but I wanted to ask if it was getting addressed. Specifically, it is the inability for Tags to be prefetched so when you try to query tags on a model, a separate query is made for every single tag. The issue I am having is even though I only have a few entries, they each have a series of tags and when I try to fetch the model data and tags using DRF it makes hundreds of queries and the response time of around 10 seconds. I have never had this issue with any other Models or serializers. Prefetch has always helped keep queries to a minimum. However, When I try to prefetch "tags" I get double the queries than when I do no prefetching on tags.   This isn't a workable solution and I was wondering if the issue has been resolved or is being worked on? Is there a possible workaround or should I not use tags?

From a Changelog in 2016 I found this reference to ClusterTaggableManager that says prefetch doesn't work on ClusterTaggableManager. Is this the reason prefetch is not working on my tags?

Are tags not meant to be used as filters when using DRF? 

https://github.com/wagtail/django-modelcluster/blob/master/CHANGELOG.txt
2.0 (22.04.2016)
Fix: prefetch_related on a ClusterTaggableManager no longer fails (but doesn't prefetch either)

Any direction you can provide would be much appreciated. Again as I wasn't sure if this is a bug or a mere limitation, I also opened up a StackOverflow question here:
https://stackoverflow.com/questions/61884503/how-do-i-query-wagtail-tags-without-it-creating-hundreds-of-queries-n1-problem

### Steps to Reproduce

```

# serializer.py////////////////////////////////////

# orderable models links
# Custom searializers to add custom display fields
class EventCollectionSerializer(serializers.ModelSerializer):
    class Meta:
         model = CategoryEventCollection
         fields = ['id', 'collection_name', 'collection_event']

class TagSerializer(serializers.ModelSerializer):
    class Meta:
         model = Tag
         fields = ['name', 'id']


# MAIN SERIALIZER PRIMARY FOCUS
class PrimaryFocusSerializer(serializers.HyperlinkedModelSerializer):

     # assign fields to custom serializer
     event_collection = EventCollectionSerializer(many=False)
     tags = TagSerializer(many=True)


class Meta:
    model = PrimaryFocus
    fields = ['id',
              'event_collection',
              'title_major',
              'synopsis',
              'level_ability',
              'page_count',
              'reading_category',
              'book_image',
              'author_first_name',
              'author_last_name',
              'author_dob',
              'tags',
              'purchase_link',
              'source',
              ]



# PRIMARY FOCUS VIEWSET//////////////////////////////////
class PrimaryFocusViewSet(viewsets.ModelViewSet):

    # This snippet just console logs how many queries
    def dispatch(self, *args, **kwargs):
        response = super().dispatch(*args, **kwargs)
        print('Queries Counted: {}'.format(len(connection.queries)))
        return response

# fetch all lexis items to setup queryset
queryset = PrimaryFocus.objects.all()

# set serializer class
serializer_class = PrimaryFocusSerializer

# optimize query with select related and prefetching
def get_queryset(self):
    return (PrimaryFocus.objects
            .select_related('event_collection',)
            .prefetch_related('tags', )
            .all()
            )
```

### Technical details

* Python version: 3.7.6
* Django version: 3.0
* Wagtail version: 2.9
* Browser version: Chrome 81

**Comment by dopry** [comment]
@Redhaus did you ever figure out how to improve tagging performance? We implemented 6 tag categories as individual relations to Pages with about 20-30 options in each and well...  it takes 960 queries and 20s to load  a page. I strip 'em out and its only 44 queries and 1s for the same page...

**Comment by chosak** [comment]
This n+1 issue exists generally with Wagtail's recommended tagging approach, independent of use of DRF or not as mentioned in the original question.

Using the [example in the docs](https://docs.wagtail.org/en/stable/reference/pages/model_recipes.html#tagging),   prefetching doesn't currently work as expected:

```py
class BlogPageTag(TaggedItemBase):
    content_object = ParentalKey('demo.BlogPage', on_delete=models.CASCADE, related_name='tagged_items')

class BlogPage(Page):
    tags = ClusterTaggableManager(through=BlogPageTag, blank=True)

pages = BlogPage.objects.prefetch_related("tags").all()
for page in pages:
    tags = page.tags.all() # this re-queries the database for each page :(
```

I've opened https://github.com/wagtail/django-modelcluster/pull/180 in an attempt to make this work more optimally.

**Comment by gasman** [comment]
Have now released django-modelcluster 6.2 which includes @chosak's prefetching fix.
